xarray/core/dataarray.py

```diff
@@ -484,7 +484,9 @@ def __setitem__(self, key, value):
         if isinstance(key, basestring):
             self.coords[key] = value
         else:
-            # xarray-style array indexing
+            # DataArray key -> Variable key
+            key = {k: v.variable if isinstance(v, DataArray) else v
```

Comment:
Do we want to enforce consistency for coordinates here? My inclination would be that we should support exactly the same keys in setitem as are valid in getitem. Ideally we should also reuse the same code. That means we should raise errors if there are multiple indexers with inconsistent alignment.

Reply:
> My inclination would be that we should support exactly the same keys in setitem as are valid in getitem. 

Reasonable. I will add a validation.

Reply:
> My inclination would be that we should support exactly the same keys in setitem as are valid in getitem. 

Reasonable. I will add a validation.